src/lib.rs

```diff
@@ -819,187 +835,197 @@ impl<L: SynthLanguage> Synthesizer<L> {
+    fn run_chunk_rule_lifting(&mut self, chunk: &[L], iter: usize) {
+        Synthesizer::add_chunk(&mut self.egraph, chunk, iter > self.params.ema_above_iter);
 
-        assert!(self.params.iters > 0);
-        for iter in 1..=self.params.iters {
-            log::info!("[[[ Iteration {} ]]]", iter);
-            let layer = self.enumerate_layer(iter);
-            let chunk_count = div_up(layer.len(), self.params.node_chunk_size);
-            let mut chunk_num = 1;
+        let mut new_candidate_eqs: EqualityMap<L> = EqualityMap::default();
+        let run_rewrites_before = Instant::now();
 
-            log::info!("Filtered layer, {} nodes remaining", layer.len());
-            log::info!("Running loop with {} chunks", chunk_count);
-            for chunk in layer.chunks(self.params.node_chunk_size) {
-                log::info!("Chunk {} / {}", chunk_num, chunk_count);
-                log::info!(
-                    "egraph n={}, e={}",
-                    self.egraph.total_size(),
-                    self.egraph.number_of_classes(),
-                );
+        // run lifting rules
+        log::info!("running lifting rules");
+        let runner = self
+            .mk_cvec_less_runner(self.egraph.clone())
+            .with_iter_limit(self.params.eqsat_iter_limit + 1); // increase iter limit by 1 to ensure saturation
+        let rewrites: Vec<&Rewrite<L, SynthAnalysis>> = self.lifting_rewrites.iter().collect();
+        let (new_egraph, found_unions) = self.run_rewrites_with_unions(rewrites, runner);
 
-                chunk_num += 1;
-                for node in chunk {
-                    if iter > self.params.ema_above_iter {
-                        let rec = node.to_recexpr(|id| self.egraph[id].data.simplest.as_ref());
-                        L::add_domain_expr(&mut self, &rec);
-                    } else {
-                        L::add_domain_node(&mut self, node.clone());
+        // These unions are candidate rewrite rules.
+        // Do not apply unions on `egraph` since we are
+        // using `new_egraph` after this point
+        for ids in found_unions.values() {
+            for win in ids.windows(2) {
+                if win[0] != win[1]
+                    && self.egraph[win[0]].data.is_extractable
+                    && self.egraph[win[1]].data.is_extractable
+                {
+                    let extract = Extractor::new(&self.egraph, ExtractableAstSize);
+                    let (_, e1) = extract.find_best(win[0]);
+                    let (_, e2) = extract.find_best(win[1]);
+                    if let Some(mut eq) = Equality::new(&e1, &e2) {
+                        if e1 != e2 &&
+                            e1.as_ref().iter().all(|x| x.is_extractable()) &&        // extractable and valid
+                            e2.as_ref().iter().all(|x| x.is_extractable())
+                        {
+                            if let ValidationResult::Valid = L::validate(self, &eq.lhs, &eq.rhs) {
```

Comment:
why this and not `is_valid_rewrite`? Why do we need is_valid_rewrite again?

Reply:
Bad naming convention. I believe that `validate()` returns whether or not a rule is valid to return to the user while `is_valid_rewrite` checks if the rule should be applied.